[#1411] Refactor AccountBalance to use Balance for transparent funds (#1570)
zcash/librustzcash: zcash_primitives/CHANGELOG.md

[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- `zcash_client_backend::AccountBalance::with_unshielded_balance_mut`

### Deprecated

> - `zcash_client_backend::AccountBalance::add_unshielded_value`

Reviewer: ```suggestion
- `zcash_client_backend::AccountBalance::add_unshielded_value`. Instead use
  `AccountBalance::with_unshielded_balance_mut` with a closure that calls the
  appropriate `add_*_value` method(s) of `Balance` on its argument. Note that
  the appropriate method(s) depend on whether the funds are spendable,
  pending change, or pending non-change. The deprecated implementation
  now assumes that they are pending non-change, which may not be correct.
```

> - `zcash_client_backend::AccountBalance::unshielded`

Reviewer: ```suggestion
- `zcash_client_backend::AccountBalance::unshielded`.
  Use `account_balance.unshielded_balance().total()` instead.
```

> ### Deprecated
> - `zcash_client_backend::AccountBalance::add_unshielded_value`
> - `zcash_client_backend::AccountBalance::unshielded`

Reviewer: `zcash_client_backend::AccountBalance::unshielded` is removed, not deprecated.

Option 1 from above:
```suggestion
### Removed
- `zcash_client_backend::AccountBalance::unshielded`. Instead use
  `account_balance.unshielded_balance().total()`.

### Deprecated
- `zcash_client_backend::AccountBalance::add_unshielded_value`. Instead use
  `AccountBalance::with_unshielded_balance_mut` with a closure that calls
  the appropriate `add_*_value` method(s) of `Balance` on its argument.
  Note that the appropriate method(s) depend on whether the funds are
  spendable, pending change, or pending non-change (previously, only the
  total unshielded value was tracked). The deprecated implementation now
  assumes that they are pending non-change, which may not be correct, and
  so all uses of this method should be checked and replaced.
```

Option 2 from above:
```suggestion
### Removed
- `zcash_client_backend::AccountBalance::unshielded`. Instead use
  `account_balance.unshielded_balance().total()`.
- `zcash_client_backend::AccountBalance::add_unshielded_value`. Instead use
  `AccountBalance::with_unshielded_balance_mut` with a closure that calls
  the appropriate `add_*_value` method(s) of `Balance` on its argument.
  Note that the appropriate method(s) depend on whether the funds are
  spendable, pending change, or pending non-change (previously, only the
  total unshielded value was tracked).
```

Author: I'm settling for option two (if possible) since it would be a better contribution overall

## [0.19.0] - 2024-10-02

### Changed
- Migrated to `zcash_address 0.6`.

> - Implemented changes to refactor AccountBalance to use Balance for transparent funds
> (see issue #1411). `AccountBalance` now has an `unshielded` value that uses Balance.
> The current implementation maintains retrocompatibility with the `unshielded` value
> represeted with a `NonNegativeAmount`. There are values that are pending to be 
> implemented such as change tracking.

Reviewer: ```suggestion
- Refactored `AccountBalance` to use `Balance` for transparent funds (issue #1411).
  `AccountBalance` now has an `unshielded` value that uses `Balance`. This does not
  currently distinguish between pending change and non-change; the pending value
  is all counted as non-change.
```
Is the intention to fix the latter issue in this PR?

Author: Yes! I'm figuring out how to do it 😅. Also I will add more tests.

Author: Update: Given how `zcash_client_backend` and the sqlite crate accounts for transparent funds, this is a bigger issue that has dependencies that are out of scope for issue #1411. I created #1592 which describes the follow-up work and (some of) the related issues that are already tracked by the [Zcashd deprecation DAG](https://zcash.github.io/developers/zcash-zcashd-deprecation-dag)

[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- `zcash_client_backend::AccountBalance::with_unshielded_balance_mut`

### Deprecated
- `zcash_client_backend::AccountBalance::unshielded`. Instead use
  `account_balance.unshielded_balance().total()`.
  
### Removed
- `zcash_client_backend::AccountBalance::add_unshielded_value`. Instead use
  `AccountBalance::with_unshielded_balance_mut` with a closure that calls
  the appropriate `add_*_value` method(s) of `Balance` on its argument.
  Note that the appropriate method(s) depend on whether the funds are
  spendable, pending change, or pending non-change (previously, only the
  total unshielded value was tracked).

> ## [0.19.0] - 2024-10-02

Reviewer: ```suggestion

## [0.19.0] - 2024-10-02
```
(There needs to be a blank line between sections in Markdown.)

### Changed
- Migrated to `zcash_address 0.6`.
- Refactored `AccountBalance` to use `Balance` for transparent funds (issue #1411).
  `AccountBalance` now has an `unshielded_balance()` that uses `Balance`. This does
  not currently distinguish between pending change and non-change; the pending value
  is all counted as non-change.

>   

Reviewer: ```suggestion
```
(just a nit)

### Changed
- MSRV is now 1.77.0.
- Refactored `AccountBalance` to use `Balance` for transparent funds (issue #1411).
  `AccountBalance` now has an `unshielded_balance()` function that returns `Balance`. 
  This does not currently distinguish between pending change and non-change; the 
  pending value is all counted as non-change. Ability to distinguish such funds is 

>   tracked by issue #1592

Reviewer: ```suggestion
  tracked by issue #1592.
```

- A new feature flag, `non-standard-fees`, has been added. This flag is now
  required in order to make use of any types or methods that enable non-standard
  fee calculation.

> - `zcash_client_backend::AccountBalance::with_unshielded_balance_mut`

Reviewer: This is still in the wrong changelog.

  - `zip317::FeeRule::non_standard` has been moved behind the `non-standard-fees`
    feature flag. Using a non-standard fee may result in a transaction that cannot
    be mined on the current `Zcash` network.
- Refactored `AccountBalance` to use `Balance` for transparent funds (issue #1411).
  `AccountBalance` now has an `unshielded_balance()` function that returns `Balance`. 
  This does not currently distinguish between pending change and non-change; the 
  pending value is all counted as non-change. Ability to distinguish such funds is 

>   tracked by issue #1592

Reviewer: This part is not part of the `WalletRead` API; it's a bug/limitation in `zcash_client_sqlite`'s implementation of that API for `WalletDb`. All other changes to this file should instead be to `zcash_client_backend/CHANGELOG.md`.

Author: Oh, that's right. Thank you for pointing it out.

    ZIP 317 fee, use `zip317::FeeRule::standard()`. To preserve the current
    behaviour, use `fixed::FeeRule::non_standard(zip317::MINIMUM_FEE)`,
    but note that this is likely to result in transactions that cannot be mined.
- `zcash_client_backend::AccountBalance::add_unshielded_value`. Instead use
  `AccountBalance::with_unshielded_balance_mut` with a closure that calls
  the appropriate `add_*_value` method(s) of `Balance` on its argument.
  Note that the appropriate method(s) depend on whether the funds are
  spendable, pending change, or pending non-change (previously, only the

>   total unshielded value was tracked).

Reviewer: This is still in the wrong changelog.
